### PR TITLE
protected route issue#28

### PR DESCRIPTION
**Pull request by seblaifa3315**
### What this PR does (required):
- I created a superset of Route and add in an authentication check.

### Any information needed to test this feature (required):
- I applied the ProtectedRoute component to the component Dashboard and Settings in App.tsx

### Any issues with the current functionality (optional):
- I'm not sure I understood correctly the ticket (issue#28). Let me know if I am on the right direction.

### CODE REVIEW

File: client/src/components/ProtectedRoute/ProtectedRoute.tsx [hunk 1]
```diff
@@ -0,0 +1,15 @@
+import { Route, RouteProps, Redirect } from 'react-router-dom';
+import { User } from '../../interface/User';
+
+interface Props extends RouteProps {
+  loggedInUser: User | null | undefined;
+}
+
+const ProtectedRoute = ({ loggedInUser, ...routeProps }: Props) => {
+  if (loggedInUser) {
```
Comment (ugur-akin) [inline client/src/components/ProtectedRoute/ProtectedRoute.tsx:9]: This lives in the context, so no need to pass it down - we can simply use the context in this component too:
```
const { loggedInUser } = useAuth();
```

File: client/src/App.tsx [hunk 2]
```diff
@@ -13,8 +13,11 @@ import { SnackBarProvider } from './context/useSnackbarContext';
 import { Navbar } from './components/Navbar/Navbar';
 import Settings from './pages/Settings/Settings';
 import NotFound from './pages/NotFound/NotFound';
+import ProtectedRoute from './components/ProtectedRoute/ProtectedRoute';
+import { useAuth } from './context/useAuthContext';
 
 function App(): JSX.Element {
+  const { loggedInUser } = useAuth();
```
Comment (ugur-akin) [inline client/src/App.tsx:20]: We can move this down to the `ProtectedRoute` component 👍🏼

**Review by ugur-akin — COMMENTED** [review]
Hey @seblaifa3315 , looks good! You're pretty much spot-on, there is just a minor context usage adjustment we can make to simplify things. Please have a look and let's check back soon 👍🏼 

Also, I wanted to bring some key points to your attention regarding how to communicate your PRs more effectively:
- Ideally, the title is descriptive and written in plain English (e.g. `Implements auth-protected routes`)
- We can mention the issue in the PR body, ideally using one of the [supported keywords for automatically linking it to your PR](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword). This should also automatically connect the issue on ZenHub so it's a huge plus for project management 🎉

To learn more, I recommend having a look at [this article](https://hugooodias.medium.com/the-anatomy-of-a-perfect-pull-request-567382bb6067) when you can spare the time 👍🏼 

Bravo, à bientôt :P

File: client/src/App.tsx [hunk 3]
```diff
@@ -26,8 +27,8 @@ function App(): JSX.Element {
               <Switch>
                 <Route exact path="/login" component={Login} />
                 <Route exact path="/signup" component={Signup} />
-                <Route exact path="/dashboard" component={Dashboard} />
-                <Route path="/profile/settings" component={Settings} />
+                <ProtectedRoute exact path="/dashboard" component={Dashboard} />
```
Comment (rajivtitus) [inline client/src/App.tsx:30]: 👍🏽

File: client/src/components/ProtectedRoute/ProtectedRoute.tsx [hunk 4]
```diff
@@ -0,0 +1,13 @@
+import { Route, RouteProps, Redirect } from 'react-router-dom';
+import { useAuth } from '../../context/useAuthContext';
+
+const ProtectedRoute = ({ ...routeProps }: RouteProps) => {
+  const { loggedInUser } = useAuth();
+
```
Comment (rajivtitus) [inline client/src/components/ProtectedRoute/ProtectedRoute.tsx:6]: Optional: Additionally, we can also display a loading spinner while we fetch the logged in user's details (for example, on page refresh we will be directed to the login page for a split second)

**Review by rajivtitus — APPROVED** [review]
LGTM! Thanks for the changes 👍🏽 
Just an optional comment for you to have a look

**Comment by seblaifa3315** [comment]
I added the loading spinner as addressed in the comments above.
Let me know if everything is correct, thanks.

**Comment by rajivtitus** [comment]
> added loading spinnner

LGTM!
